### PR TITLE
Move Workflow functionalities to separated service class

### PR DESCRIPTION
**Pull request by Beacze**
- change task status
- correction of tasks
- tests

### CODE REVIEW

File: Kitodo/src/main/java/de/sub/goobi/helper/BatchStepHelper.java [hunk 1]
```diff
@@ -220,10 +217,14 @@ private void saveStep() {
...
     public String reportProblemForSingle() {
-
         this.myDav.uploadFromHome(this.currentStep.getProcess());
-        reportProblem();
-        this.problemMessage = "";
+        serviceManager.getWorkflowService().setProblem(getProblem());
+        try {
+            this.currentStep = serviceManager.getWorkflowService().reportProblem(this.currentStep, this.problemTask);
+        } catch (DataException e) {
+            logger.error("Problem couldn't be reported: " + e);
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/de/sub/goobi/helper/BatchStepHelper.java]: 2 things:
1) If a `DataException` is thrown is it a good idea to go on?
2) Did this kind of error logging work with casting exception object to string message?

File: Kitodo/src/main/java/de/sub/goobi/helper/BatchStepHelper.java [hunk 2]
```diff
@@ -237,74 +238,20 @@ public String reportProblemForAll() {
         for (Task s : this.steps) {
             this.currentStep = s;
             this.myDav.uploadFromHome(this.currentStep.getProcess());
-            reportProblem();
+            serviceManager.getWorkflowService().setProblem(getProblem());
+            try {
+                setCurrentStep(serviceManager.getWorkflowService().reportProblem(this.currentStep, this.problemTask));
+            } catch (DataException e) {
+                logger.error("Problem couldn't be reported: " + e);
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/de/sub/goobi/helper/BatchStepHelper.java]: 2 things:
1) If a `DataException` is thrown is it a good idea to go on?
2) Did this kind of error logging work with casting exception object to string message?

File: Kitodo/src/main/java/de/sub/goobi/helper/BatchStepHelper.java [hunk 3]
```diff
@@ -343,9 +290,14 @@ private void reportProblem() {
      */
     public String solveProblemForSingle() {
         try {
-            solveProblem();
+            serviceManager.getWorkflowService().setSolution(getSolution());
+            try {
+                setCurrentStep(serviceManager.getWorkflowService().solveProblem(this.currentStep, this.solutionTask));
+            } catch (DataException e) {
+                logger.error("Problem couldn't be solved: " + e);
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/de/sub/goobi/helper/BatchStepHelper.java]: 2 things:
1) If a `DataException` is thrown is it a good idea to go on?
2) Did this kind of error logging work with casting exception object to string message?

File: Kitodo/src/main/java/de/sub/goobi/helper/BatchStepHelper.java [hunk 4]
```diff
@@ -365,10 +317,15 @@ public String solveProblemForAll() {
         try {
             for (Task s : this.steps) {
                 this.currentStep = s;
-                solveProblem();
+                serviceManager.getWorkflowService().setSolution(getSolution());
+                try {
+                    setCurrentStep(serviceManager.getWorkflowService().solveProblem(this.currentStep, this.solutionTask));
+                } catch (DataException e) {
+                    logger.error("Problem couldn't be solved: " + e);
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/de/sub/goobi/helper/BatchStepHelper.java]: 2 things:
1) If a `DataException` is thrown is it a good idea to go on?
2) Did this kind of error logging work with casting exception object to string message?

File: Kitodo/src/main/java/org/kitodo/services/workflow/WorkflowService.java [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+     * @param task
+     *            object
+     * @return closed Task
+     */
+    public Task closeTaskByUser(Task task) throws DataException, IOException {
+
+        // if step allows writing of images, then count all images here
+        if (task.isTypeImagesWrite()) {
+            try {
+                // this.mySchritt.getProzess().setSortHelperImages(
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/services/workflow/WorkflowService.java]: Please remove out commented code. This code is commented out even in 2.x branch.

File: Kitodo/src/main/java/org/kitodo/services/workflow/WorkflowService.java [hunk 6]
```diff
@@ -0,0 +1,726 @@
...
+
+        updateProcessSortHelperStatus(currentTask.getProcess());
+
+        this.problem.setMessage("");
+        this.problem.setId(0);
+        return currentTask;
+    }
+
+    /**
+     * THis one is taken out of BatchStepHelper.
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/services/workflow/WorkflowService.java]: Typo in `THis` ;)

File: Kitodo/src/main/java/org/kitodo/workflow/Solution.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+ *
+ * It is licensed under GNU General Public License version 3 or later.
+ *
+ * For the full copyright and license information, please read the
+ * GPL3-License.txt file that was distributed with this source code.
+ */
+
+package org.kitodo.workflow;
+
+public class Solution extends Message{
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/workflow/Solution.java]: Missing space between `Message` and `{`.

File: Kitodo/src/main/java/de/sub/goobi/helper/BatchStepHelper.java (hunk 1 above)
Comment (Beacze) [reply Kitodo/src/main/java/de/sub/goobi/helper/BatchStepHelper.java]: 1. Partially yes, partially no. Method saveStep shouldn't be called but it shoul give back the list of tasks. 
2. Yes, the whole exception will be printed out.

File: Kitodo/src/main/java/de/sub/goobi/helper/BatchStepHelper.java (hunk 2 above)
Comment (Beacze) [reply Kitodo/src/main/java/de/sub/goobi/helper/BatchStepHelper.java]: Look previous.

File: Kitodo/src/main/java/de/sub/goobi/helper/BatchStepHelper.java (hunk 3 above)
Comment (Beacze) [reply Kitodo/src/main/java/de/sub/goobi/helper/BatchStepHelper.java]: Look previous.

File: Kitodo/src/main/java/de/sub/goobi/helper/BatchStepHelper.java (hunk 4 above)
Comment (Beacze) [reply Kitodo/src/main/java/de/sub/goobi/helper/BatchStepHelper.java]: Look previous.

File: Kitodo/src/main/java/org/kitodo/services/workflow/WorkflowService.java (hunk 5 above)
Comment (Beacze) [reply Kitodo/src/main/java/org/kitodo/services/workflow/WorkflowService.java]: Done.

File: Kitodo/src/main/java/org/kitodo/services/workflow/WorkflowService.java (hunk 6 above)
Comment (Beacze) [reply Kitodo/src/main/java/org/kitodo/services/workflow/WorkflowService.java]: Done.

File: Kitodo/src/main/java/org/kitodo/workflow/Solution.java (hunk 7 above)
Comment (Beacze) [reply Kitodo/src/main/java/org/kitodo/workflow/Solution.java]: Done.

**Pull request merged**